File: frontend/src/services/hrService.ts
import { apiClient } from './authService'

export interface Employee {
  id: number
  first_name: string
  last_name: string
  email: string
  phone?: string
  department?: string
  title?: string
  status: string
  created_at: string
  updated_at: string
}

export interface Attendance {
  id: number
  employee_id: number
  date: string
  check_in?: string
  check_out?: string
  notes?: string
  created_at: string
  updated_at: string
}

export interface Payroll {
  id: number
  employee_id: number
  period_start: string
  period_end: string
  base_salary: number
  allowances: number
  deductions: number
  net_pay: number
  status: string
  created_at: string
  updated_at: string
}

export interface Leave {
  id: number
  employee_id: number
  start_date: string
  end_date: string
  type: string
  status: string
  reason?: string
  created_at: string
  updated_at: string
}

class HRService {
  async getEmployees(params?: { skip?: number; limit?: number; q?: string }) {
    const { data } = await apiClient.get('/hr/employees', { params })
    return data
  }

  async createEmployee(payload: Partial<Employee>) {
    const { data } = await apiClient.post('/hr/employees', payload)
    return data
  }

  async getAttendance(params?: { employee_id?: number; skip?: number; limit?: number }) {
    const { data } = await apiClient.get('/hr/attendance', { params })
    return data
  }

  async createAttendance(payload: Partial<Attendance>) {
    const { data } = await apiClient.post('/hr/attendance', payload)
    return data
  }

  async getPayrolls(params?: { employee_id?: number; skip?: number; limit?: number }) {
    const { data } = await apiClient.get('/hr/payrolls', { params })
    return data
  }

  async createPayroll(payload: Partial<Payroll>) {
    const { data } = await apiClient.post('/hr/payrolls', payload)
    return data
  }

  async getLeaves(params?: { employee_id?: number; skip?: number; limit?: number }) {
    const { data } = await apiClient.get('/hr/leaves', { params })
    return data
  }

  async createLeave(payload: Partial<Leave>) {
    const { data } = await apiClient.post('/hr/leaves', payload)
    return data
  }
}

export const hrService = new HRService()

File: backend/app/services/hr_service.py
from typing import List, Optional
from sqlalchemy.orm import Session
from sqlalchemy import desc
from datetime import datetime

from ..models.hr import Employee, Attendance, Payroll, Leave
from ..schemas.hr import (
    EmployeeCreate,
    EmployeeUpdate,
    AttendanceCreate,
    PayrollCreate,
    LeaveCreate,
)


class HRService:
    def __init__(self, db: Session):
        self.db = db

    # Employees
    def create_employee(self, data: EmployeeCreate) -> Employee:
        employee = Employee(**data.dict())
        self.db.add(employee)
        self.db.commit()
        self.db.refresh(employee)
        return employee

    def update_employee(self, employee_id: int, data: EmployeeUpdate) -> Optional[Employee]:
        employee = self.db.query(Employee).filter(Employee.id == employee_id).first()
        if not employee:
            return None
        for field, value in data.dict(exclude_unset=True).items():
            setattr(employee, field, value)
        self.db.commit()
        self.db.refresh(employee)
        return employee

    def list_employees(self, skip: int = 0, limit: int = 10, q: Optional[str] = None) -> List[Employee]:
        query = self.db.query(Employee)
        if q:
            like = f"%{q}%"
            query = query.filter((Employee.first_name.ilike(like)) | (Employee.last_name.ilike(like)) | (Employee.email.ilike(like)))
        return query.order_by(desc(Employee.created_at)).offset(skip).limit(limit).all()

    def count_employees(self, q: Optional[str] = None) -> int:
        query = self.db.query(Employee)
        if q:
            like = f"%{q}%"
            query = query.filter((Employee.first_name.ilike(like)) | (Employee.last_name.ilike(like)) | (Employee.email.ilike(like)))
        return query.count()

    # Attendance
    def create_attendance(self, data: AttendanceCreate) -> Attendance:
        attendance = Attendance(**data.dict())
        self.db.add(attendance)
        self.db.commit()
        self.db.refresh(attendance)
        return attendance

    def list_attendance(self, employee_id: Optional[int] = None, skip: int = 0, limit: int = 10) -> List[Attendance]:
        query = self.db.query(Attendance)
        if employee_id:
            query = query.filter(Attendance.employee_id == employee_id)
        return query.order_by(desc(Attendance.date)).offset(skip).limit(limit).all()

    def count_attendance(self, employee_id: Optional[int] = None) -> int:
        query = self.db.query(Attendance)
        if employee_id:
            query = query.filter(Attendance.employee_id == employee_id)
        return query.count()

    # Payroll
    def create_payroll(self, data: PayrollCreate) -> Payroll:
        payload = data.dict()
        if payload.get("net_pay") is None:
            payload["net_pay"] = payload.get("base_salary", 0) + payload.get("allowances", 0) - payload.get("deductions", 0)
        payroll = Payroll(**payload)
        self.db.add(payroll)
        self.db.commit()
        self.db.refresh(payroll)
        return payroll

    def list_payrolls(self, employee_id: Optional[int] = None, skip: int = 0, limit: int = 10) -> List[Payroll]:
        query = self.db.query(Payroll)
        if employee_id:
            query = query.filter(Payroll.employee_id == employee_id)
        return query.order_by(desc(Payroll.period_start)).offset(skip).limit(limit).all()

    def count_payrolls(self, employee_id: Optional[int] = None) -> int:
        query = self.db.query(Payroll)
        if employee_id:
            query = query.filter(Payroll.employee_id == employee_id)
        return query.count()

    # Leave
    def create_leave(self, data: LeaveCreate) -> Leave:
        leave = Leave(**data.dict())
        self.db.add(leave)
        self.db.commit()
        self.db.refresh(leave)
        return leave

    def list_leaves(self, employee_id: Optional[int] = None, skip: int = 0, limit: int = 10) -> List[Leave]:
        query = self.db.query(Leave)
        if employee_id:
            query = query.filter(Leave.employee_id == employee_id)
        return query.order_by(desc(Leave.start_date)).offset(skip).limit(limit).all()

    def count_leaves(self, employee_id: Optional[int] = None) -> int:
        query = self.db.query(Leave)
        if employee_id:
            query = query.filter(Leave.employee_id == employee_id)
        return query.count()


